### DISCUSSION TITLE
Error starting up on Unraid

### DISCUSSION THREAD

**Discussion by ryphez** [philosowaffle/peloton-to-garmin#286] [Q&A]
I am not sure when the problem started occuring unfortunately, but I had my setup working on Unraid with the following configuration pulled from DockerHub image.

New Container Path:` /config -> /mnt/user/appdata/peloton-to-garmin/`
Where */app/configuration.local.json exists

I have all of a sudden started getting this error on startup which makes me feel like the config loading method I had working previously no longer works given the config never changed.

`today at 9:56:01 PMWelcome! P2G is starting up...
today at 9:56:01 PM[21:56:01 ERR] Peloton Email required, check your configuration Peloton.Email is set.
today at 9:56:01 PM[21:56:01 ERR] Configuration invalid
today at 9:56:01 PMSystem.ArgumentException: Peloton Email must be set. (Parameter 'Email')
today at 9:56:01 PM   at Peloton.PelotonService.ValidateConfig(Peloton config) in /build/src/Peloton/PelotonService.cs:line 51
today at 9:56:01 PM   at Program.<>c.<<Main>$>b__0_5(IServiceProvider serviceProvider)`

Any help with path mapping would be great to help me debug what changed here!

**Comment by philosowaffle** [comment]
I'm not familiar with unraid's way of doing things, can you elaborate a bit more on what this configuration means:

> New Container Path: /config -> /mnt/user/appdata/peloton-to-garmin/
  Where */app/configuration.local.json exists

In docker compose or the traditional docker run command, this would look something like:

```
/path/on/myComputer/configuration.local.json:/app/configuration.local.json
```

Where the left hand side of the `:` is the path on your local computer, this can be any path so long as it points to a valid `configuration.local.json` with read permissions.  And the right hand side of the `:` denotes the path within the P2G container itself, this path is fixed and cannot be modified.

**Comment by ryphez** [comment]
You can map Paths to directories in unraid, not to a particular file. Did the docker compose use to have a "/config"? I'm trying to understand why it worked before and now is breaking

**Reply by philosowaffle** [reply]
I looked back in time to when the docker-compose was originally documented, just to make sure I hadn't accidentally changed something, and [seems it has always been under the /app dir](https://github.com/philosowaffle/peloton-to-garmin/commit/67ee6f574335227ce525360d74a56f1015b7cf67).

**Reply by ryphez** [reply]
Ya I checked the old code as well and nothing stood out. I'll leave this open for now if anyone else has done this on Unraid, but will work to mess with it in a week after vacation! Thanks for the awesome project!

**Reply by philosowaffle** [reply]
I know @pgray007 is also running on Unraid, so maybe they'll have some insight, sorry I couldn't be more help.

**Comment by lowgatelabs** [comment]
I only have 3 variables setup in unraid (and am supplying my values for reference). I do remember having some problems getting it to find the config file:

TZ: Ameriza/New_Your
working: (blank, previously populated but it caused a write lock error)
output: /mnt/user/appdata/peloton-to-garmin/output/
config: /mnt/user/appdata/peloton-to-garmin/configuration.local.json

It's been a while since I mucked with the docker settings, but this seems to have kept me up and running for several months.

**Reply by ryphez** [reply]
Back from vacation and trying this. Do you have those 3 items as paths or variables?

**Reply by lowgatelabs** [reply]
I believe they are variables. I'm working in unpaid so it's not always easy to tell, but here's the docker run command that ultimately gets executed:

root@localhost:# /usr/local/emhttp/plugins/dynamix.docker.manager/scripts/docker run -d --name='peloton-to-garmin' --net='bridge' -e TZ="America/New_York" -e HOST_OS="Unraid" -e HOST_HOSTNAME="hostname" -e HOST_CONTAINERNAME="peloton-to-garmin" -e 'TZ'='America/New_York' -l net.unraid.docker.managed=dockerman -l net.unraid.docker.icon='/plugins/dynamix.docker.manager/images/question.png' -v '':'/app/working':'rw' -v '/mnt/user/appdata/peloton-to-garmin/output/':'/app/output':'rw' -v '/mnt/user/appdata/peloton-to-garmin/configuration.local.json':'/app/configuration.local.json':'rw' 'philosowaffle/peloton-to-garmin:stable' 
dockerIDremoved

I don't get the error anymore, but it still seems the app downloads the last several workouts and attempts to upload them to Garmin connect after every sleep cycle (yes, I've been doing a lot of stretching lately):

Welcome! P2G is starting up...
[06:21:38 INF] App Version: 3.0.2
[06:21:38 INF] Operating System: Unix 5.15.46.0
[06:21:38 INF] DotNet Runtime: 6.0.8
[06:21:38 INF] Application started. Press Ctrl+C to shut down.
[06:21:38 INF] Hosting environment: Production
[06:21:38 INF] Content root path: /app
[06:21:39 INF] [Fit] Encoded file /app/working/ef30adbda3ff4e29bbb646e61b9541f9_10_min_Full_Body_Stretch_with_Jermaine_Johnson.Fit
[06:21:39 INF] [Fit] Encoded file /app/working/a944431acf954eff92f881fde32b4c26_10_min_Tropical_House_Full_Body_Stretch_with_Ben_Alldis.Fit
[06:21:39 INF] [Fit] Encoded file /app/working/5ebccf9b3c3841b0b9ba7262de83ae82_10_min_Post-Run_Stretch_with_Jess_Sims.Fit
[06:21:39 INF] [Fit] Encoded file /app/working/59397f8f21cf4dffae1da9245bc18364_10_min_Post-Run_Stretch_with_Jon_Hosking.Fit
[06:21:39 INF] [Fit] Encoded file /app/working/ddf397d5b8bb4706aa2240d1c7f92a16_10_min_Post-Run_Stretch_with_Jermaine_Johnson.Fit
[06:21:45 INF] Uploading to Garmin: /app/working/upload/a944431acf954eff92f881fde32b4c26_10_min_Tropical_House_Full_Body_Stretch_with_Ben_Alldis.Fit
[06:21:45 INF] Activity already uploaded a944431acf954eff92f881fde32b4c26_10_min_Tropical_House_Full_Body_Stretch_with_Ben_Alldis.Fit
[06:21:45 INF] Rate limiting, upload will continue after 3 seconds...
[06:21:48 INF] Uploading to Garmin: /app/working/upload/5ebccf9b3c3841b0b9ba7262de83ae82_10_min_Post-Run_Stretch_with_Jess_Sims.Fit
[06:21:49 INF] Activity already uploaded 5ebccf9b3c3841b0b9ba7262de83ae82_10_min_Post-Run_Stretch_with_Jess_Sims.Fit
[06:21:49 INF] Rate limiting, upload will continue after 4 seconds...
[06:21:53 INF] Uploading to Garmin: /app/working/upload/ef30adbda3ff4e29bbb646e61b9541f9_10_min_Full_Body_Stretch_with_Jermaine_Johnson.Fit
[06:21:54 INF] Activity already uploaded ef30adbda3ff4e29bbb646e61b9541f9_10_min_Full_Body_Stretch_with_Jermaine_Johnson.Fit
[06:21:54 INF] Rate limiting, upload will continue after 2 seconds...
[06:21:56 INF] Uploading to Garmin: /app/working/upload/ddf397d5b8bb4706aa2240d1c7f92a16_10_min_Post-Run_Stretch_with_Jermaine_Johnson.Fit
[06:21:56 INF] Activity already uploaded ddf397d5b8bb4706aa2240d1c7f92a16_10_min_Post-Run_Stretch_with_Jermaine_Johnson.Fit
[06:21:56 INF] Rate limiting, upload will continue after 1 seconds...
[06:21:57 INF] Uploading to Garmin: /app/working/upload/59397f8f21cf4dffae1da9245bc18364_10_min_Post-Run_Stretch_with_Jon_Hosking.Fit
[06:21:57 INF] Activity already uploaded 59397f8f21cf4dffae1da9245bc18364_10_min_Post-Run_Stretch_with_Jon_Hosking.Fit
[06:21:57 INF] Rate limiting, upload will continue after 4 seconds...
[06:22:02 INF] Upload complete.
[06:22:02 INF] Done
[06:22:02 INF] Sleeping for 7200 seconds...

**Reply by ryphez** [reply]
I'm able to get it to upload items to Garmin now and fetch workouts from Peloton using Docker Compose instead. I am getting an odd initial stack track though @philosowaffle 

> today at 8:55:45 PMWelcome! P2G is starting up...
> today at 8:55:45 PM[20:55:45 INF] App Version: 3.0.3
> today at 8:55:45 PM[20:55:45 INF] Operating System: Unix 5.19.3.0
> today at 8:55:45 PM[20:55:45 INF] DotNet Runtime: 6.0.8
> today at 8:55:46 PM[20:55:46 ERR] Http Call Failed. OK {...[removed]...}
> today at 8:55:46 PM[20:55:46 ERR] Failed to fetch UserData from Peloton. FTP info may be missing for certain non-class workout types (Just Ride).
> today at 8:55:46 PMFlurl.Http.FlurlParsingException: Response could not be deserialized to JSON: GET https://api.onepeloton.com/api/me
> today at 8:55:46 PM ---> Newtonsoft.Json.JsonSerializationException: Error converting value {null} to type 'Common.Dto.Peloton.CyclingFtpSource'. Path 'cycling_ftp_source', line 1, position 1073.
> today at 8:55:46 PM ---> System.ArgumentNullException: Value cannot be null. (Parameter 'value')
> today at 8:55:46 PM   at System.Enum.ToObject(Type enumType, Object value)
> today at 8:55:46 PM   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)
> today at 8:55:46 PM   --- End of inner exception stack trace ---
> today at 8:55:46 PM   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)
> today at 8:55:46 PM   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
> today at 8:55:46 PM   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
> today at 8:55:46 PM   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
> today at 8:55:46 PM   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
> today at 8:55:46 PM   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
> today at 8:55:46 PM   at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader reader, Type objectType)
> today at 8:55:46 PM   at Newtonsoft.Json.JsonSerializer.Deserialize[T](JsonReader reader)
> today at 8:55:46 PM   at Flurl.Http.Configuration.NewtonsoftJsonSerializer.Deserialize[T](Stream stream)
> today at 8:55:46 PM   at Flurl.Http.FlurlResponse.GetJsonAsync[T]()
> today at 8:55:46 PM   --- End of inner exception stack trace ---
> today at 8:55:46 PM   at Flurl.Http.FlurlRequest.HandleExceptionAsync(FlurlCall call, Exception ex, CancellationToken token)
> today at 8:55:46 PM   at Flurl.Http.FlurlResponse.GetJsonAsync[T]()
> today at 8:55:46 PM   at Flurl.Http.ResponseExtensions.ReceiveJson[T](Task`1 response)
> today at 8:55:46 PM   at Peloton.PelotonService.GetUserDataAsync() in /build/src/Peloton/PelotonService.cs:line 67
> today at 8:55:46 PM   at Sync.SyncService.SyncAsync(ICollection`1 workoutIds, ICollection`1 exclude) in /build/src/Sync/SyncService.cs:line 134
